### PR TITLE
[c10d] Fix coalescing manager when device is set

### LINKED ISSUE #122807: [c10d][NCCL] _coalescing_manager does not produce proper work handle
### 🐛 Describe the bug

The list of work objects (`works`) is empty:
```python
    with dist._coalescing_manager(group=pg_nccl, device=device, async_ops=True) as cm:
        dist.all_reduce(a)
        dist.all_reduce(b)
    print(len(cm.works)) # prints 0
    cm.wait()
```
In this situation `cm.wait()` becomes a no-op, which causes red-before-write issues whenever `_coalescing_manager` is used.

### Versions

PyTorch version: 2.4.0.dev20240326+cu121
```[tasklist]
### Tasks
- [ ] https://github.com/pytorch/pytorch/pull/122651
- [ ] https://github.com/pytorch/pytorch/pull/122849
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### PR DESCRIPTION
**Pull request by kwen2501**
Fixes #122807

Previously, if user passes in `device`, the following line would overwrite the `work` created by the Fast-Path implementation.
https://github.com/pytorch/pytorch/blob/b27ee6548d1a07b8e83570ef7a28f18d88f28e24/torch/distributed/distributed_c10d.py#L2037-L2039

Fast-Path implementation:
https://github.com/pytorch/pytorch/blob/b27ee6548d1a07b8e83570ef7a28f18d88f28e24/torch/distributed/distributed_c10d.py#L2008-L2014

This becomes an issue when the Fast-Path implementation creates the real work, while `group._end_coalescing()` creates an empty one.

The solution is to check whether we've already have a solid work. If there is, do not overwrite.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Comment by kwen2501** [comment]
@Aidyn-A

**Comment by Aidyn-A** [comment]
Hmm, may be it resolves the reproducer I proposed in #122807 but it does not fix the main issue of `wait` as `group._end_coalescing(device)` returns `None`.

**Comment by kwen2501** [comment]
Hmm, I think that wouldn't matter anymore because we are ignoring the work returned by `group._end_coalescing(device)`

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -2033,15 +2033,17 @@ def _coalescing_manager(
                 f"Coalescing manager does not support fast-path coalescing of {op0}, "
                 f"yet {op0} is still recorded in op list. This is an internal error of c10d."
             )
+        cm.append(work)
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:2036]: And we are putting the real work into `cm` by this line.

**Pull request closed without merging**